### PR TITLE
Position Control

### PR DESCRIPTION
**Pull request by eSqadron**
Position Control:
1. Actual spinning direction calculation
2. Bidirectionality
3. Precision improvement
4. Control over precision moved to KConfig

### CODE REVIEW

File: BrushedMotorControllerFirmware/include/driver.c [hunk 1]
```diff
@@ -68,11 +68,13 @@ struct DriverChannel {
 
 	const uint32_t max_pos;
 
-	void (*enc_callback_ptr)(const struct device *dev, struct gpio_callback *cb, uint32_t pins);
+	bool prev_val_enc[2];
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c]: magic number

File: BrushedMotorControllerFirmware/include/driver.c [hunk 2]
```diff
@@ -353,11 +379,49 @@ return_codes_t motor_off(enum ChannelNumber chnl)
...
+		ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P0]), 0);
+		if (ret != 0) {
+			return ERR_UNABLE_TO_SET_GPIO;
+		}
+		ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P1]), 1);
+		if (ret != 0) {
+			return ERR_UNABLE_TO_SET_GPIO;
+		}
+	}
+	return SUCCESS;
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c]:
```suggestion
	// TODO, combine if and else in some smart way, as they are very similar
	if(direction == STOPPED){
		return SUCCESS; // or other error?
	}
	ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P0]), direction == FORWARD);
	if (ret != 0) {
		return ERR_UNABLE_TO_SET_GPIO;
	}
	ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P1]), direction != FORWARD);
	if (ret != 0) {
		return ERR_UNABLE_TO_SET_GPIO;
	}
	return SUCCESS;
```

File: BrushedMotorControllerFirmware/include/shell.c [hunk 3]
```diff
@@ -442,6 +442,31 @@ static int cmd_template_size(const struct shell *shell, size_t argc, char *argv[
...
+
+	if (dir == FORWARD) {
+		shell_fprintf(shell, SHELL_INFO, "Motor is spinning forward!\n");
+	} else if (dir == BACKWARD) {
+		shell_fprintf(shell, SHELL_INFO, "Motor is spinning backward!\n");
+	} else if (dir == STOPPED) {
+		shell_fprintf(shell, SHELL_INFO, "Motor is stationary!\n");
+	} else {
+		shell_fprintf(shell, SHELL_ERROR, "Unknown direction value! - %d\n", dir);
+	}
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/shell.c:464]: convert into switch case.

File: BrushedMotorControllerFirmware/include/driver.c [hunk 4]
```diff
@@ -191,29 +212,45 @@ static void continuous_calculation_timer_handler(struct k_timer *dummy)
...
+	}
 }
-#if (CONFIG_SUPPORTED_CHANNEL_NUMBER > 1)
-static void enc_callback_ch1(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+
+static void enc_callback_wrapper(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
 {
-	enc_callback(CH1);
+	for (int channel = 0; channel < CONFIG_SUPPORTED_CHANNEL_NUMBER; ++channel) {
+		for (int pin = 0; pin < 2; ++pin) {
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c]: magic number

File: BrushedMotorControllerFirmware/include/shell.c (hunk 3 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/shell.c:464]: Already done in future PR

Comment (ghost) [reply BrushedMotorControllerFirmware/include/shell.c:464]: will come in later PR

File: BrushedMotorControllerFirmware/include/driver.c (hunk 4 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/driver.c]: fixed

File: BrushedMotorControllerFirmware/include/driver.c (hunk 2 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/driver.c]: fixed

**Review by ghost — APPROVED** [review]

**Pull request merged**